jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -0,0 +1,407 @@
+
+        advantages, targets = _calculate_gae(traj_batch, last_val)
+
+        def _update_epoch(update_state: Tuple, unused: Any) -> Tuple:
+            """Update the network for a single epoch."""
+
+            def _update_minibatch(train_state: Tuple, batch_info: Tuple) -> Tuple:
+                """Update the network for a single minibatch."""
+                params, opt_state = train_state
+                traj_batch, advantages, targets = batch_info
+
+                def _loss_fn(
+                    params: FrozenDict,
+                    opt_state: OptState,
+                    traj_batch: Transition,
+                    gae: chex.Array,
+                    targets: chex.Array,
+                ) -> Tuple:
+                    """Calculate the loss."""
+                    # RERUN NETWORK
+                    actor_policy, value = apply_fn(params, traj_batch.obs)
+                    log_prob = actor_policy.log_prob(traj_batch.action)
+
+                    # CALCULATE VALUE LOSS
+                    value_pred_clipped = traj_batch.value + (
+                        value - traj_batch.value
+                    ).clip(-config["CLIP_EPS"], config["CLIP_EPS"])
+                    value_losses = jnp.square(value - targets)
+                    value_losses_clipped = jnp.square(value_pred_clipped - targets)
+                    value_loss = (
+                        0.5 * jnp.maximum(value_losses, value_losses_clipped).mean()
+                    )
+
+                    # CALCULATE ACTOR LOSS
+                    ratio = jnp.exp(log_prob - traj_batch.log_prob)
+                    gae = (gae - gae.mean()) / (gae.std() + 1e-8)
+                    loss_actor1 = ratio * gae
+                    loss_actor2 = (
+                        jnp.clip(
+                            ratio,
+                            1.0 - config["CLIP_EPS"],
+                            1.0 + config["CLIP_EPS"],
+                        )
+                        * gae
+                    )
+                    loss_actor = -jnp.minimum(loss_actor1, loss_actor2)
+                    loss_actor = loss_actor.mean()
+                    entropy = actor_policy.entropy().mean()
+
+                    total_loss = (
+                        loss_actor
+                        + config["VF_COEF"] * value_loss
+                        - config["ENT_COEF"] * entropy
+                    )
+                    return total_loss, (value_loss, loss_actor, entropy)
+
+                grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+                total_loss, grads = grad_fn(
+                    params, opt_state, traj_batch, advantages, targets
+                )
+
+                grads, total_loss = jax.lax.pmean(
+                    (grads, total_loss), axis_name="batch"
+                )
+                grads, total_loss = jax.lax.pmean(
+                    (grads, total_loss), axis_name="device"
+                )
```

Comment:
Is `pmean`ing twice here correct, if you have a device axis then don't you want to just mean over that axis?

Reply:
I am not sure about this one maybe @RuanJohn or @DriesSmit  can have a better answer for that

Reply:
I got this from this [Anakin demo notebook](https://colab.research.google.com/drive/1974D-qP17fd5mLxy6QZv-ic4yxlPJp-G?usp=sharing). The reason for `pmean`-ing twice here is because we are `vmap`-ing over the vectorised updates as well and broadcasting the parameters to both the vectorised update functions and the devices.